See through ref pointer indirection when optimizing stack locals

333fred opened the pull request:
Fixes https://github.com/dotnet/roslyn/issues/53113. Previously, when we did stack local optimization, we'd erase the assignment (including the context of the assignment being by-ref), leaving just the pointer indirection. This is incorrect, and resulted in us generating a ldind where we shouldn't. This fixes the bug by tracking whether we're currently optimizing a stack-scheduled assignment, and seeing through the pointer indirection in that case.

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenRefConditionalOperatorTests.cs:
```diff
@@ -1202,5 +1202,90 @@ static void Main()
+}
+";
+
+            var comp = CreateCompilation(code, options: TestOptions.UnsafeReleaseExe);
```
>UnsafeReleaseExe [](http://example.com/codeflow?start=68&length=16)

We should test this case with DEBUG as well.

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenRefLocalTests.cs:
```diff
@@ -3410,5 +3410,137 @@ public static void Main()
+}
+";
+
+            var comp = CreateCompilation(code, options: TestOptions.UnsafeReleaseExe);
```
>UnsafeReleaseExe [](http://example.com/codeflow?start=68&length=16)

We should test with DEBUG as well.

cston commented:
>             if (isLast)

Shouldn't we reset `_isStackScheduledRefAssignment` here?

---
Refers to: src/Compilers/CSharp/Portable/CodeGen/Optimizer.cs:2096 in cfd9cf5. [](commit_id = cfd9cf5dca8e2366cb76bc609c275fc48bf441fc, deletion_comment = False)

333fred commented:
@cston thanks for the initial review. I've updated the implementation a bit, let me know if you want walk through it at all.

cston commented on src/Compilers/CSharp/Portable/CodeGen/Optimizer.cs:
```diff
@@ -2071,8 +2082,12 @@ public override BoundNode VisitAssignmentOperator(BoundAssignmentOperator node)
+
             // visit right
             var right = (BoundExpression)Visit(node.Right);
+            _isStackScheduledRefAssignment = false;
```
>false [](http://example.com/codeflow?start=45&length=5)

Should this reset to the previous value rather than `false`? (Is there a test we can write where that makes a difference?)

333fred replied:
No, I asserted that it was false above as well. `Visit` should be saving and resetting the value unless the thing being visited is an indirection or conditional operator, so we should never be observing a non-false value here.

cston replied:
Ok, thanks.

Could we save and restore the value anyway, for clarity, and to avoid a bug later if there are additional cases where the save and restore in `Visit` is not sufficient?

333fred replied:
I suppose I could, but I'll keep the assertion anyway. If we come across a case like that we definitely need to understand the semantics more completely.

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenRefConditionalOperatorTests.cs:
```diff
@@ -1263,7 +1263,7 @@ .maxstack  1
             {
                 var comp = CreateCompilation(code, options: options);
-                var verifier = CompileAndVerify(comp, expectedOutput: "run");
+                var verifier = CompileAndVerify(comp, expectedOutput: "run", verify: Verification.Fails);
```
I think we typically use `Verification.Skipped` when valid code is not handled by the verifier. `Verification.Fails` seems to imply a bug.

333fred replied:
Other ref local code used `Fails`, which is why I defaulted to that here.

333fred commented:
@dotnet/roslyn-compiler for a second review.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -342,6 +342,8 @@
     <!-- Non-null type is required for this node kind -->
     <Field Name="Type" Type="TypeSymbol" Override="true" Null="disallow"/>
     <Field Name="Operand" Type="BoundExpression"/>
+    <!-- Used by emit to determine whether this pointer indirection is by value or by ref -->
```
><!-- Used by emit to determine whether this pointer indirection is by value or by ref --> [](http://example.com/codeflow?start=4&length=89)

Consider including information about how to interpret the values.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -342,6 +342,8 @@
     <Field Name="Operand" Type="BoundExpression"/>
+    <!-- Used by emit to determine whether this pointer indirection is by value or by ref -->
+    <Field Name="IsAssignmentAddress" Type="bool"/>
   </Node>
```
I am not sure why do we need to calculate this value in Optimizer and limit that just to an assignment context. It feels like we can determine if a BoundPointerIndirectionOperator node is actually a dereference during an initial binding. If it is a target of a ```ref``` expression, it is not a dereference. If we capture this information during initial binding, the implementation wouldn't need to worry about other `ref` propagating scenarios like conditional expression, switch expression, etc.

333fred replied:
I like this refactoring, and went ahead and implemented it.

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs:
```diff
@@ -675,8 +675,16 @@ private void EmitAddressOfExpression(BoundAddressOfOperator expression, bool use
-            EmitLoadIndirect(expression.Type, expression.Syntax);
+            if (expression.IsAssignmentAddress)
+            {
+                EmitAddress(expression, AddressKind.Writeable);
```
>EmitAddress(expression, AddressKind.Writeable); [](http://example.com/codeflow?start=16&length=47)

It looks like we can simply do ```EmitExpression(expression.Operand, used: true);``` and share that across both branches.

AlekseyTs commented:
Done with review pass (commit 5)

alrz commented:
I imagine this will also avoids the local being stacked in a ref assignment like this: `outParam1 = outParam2 = null;`, is that correct?

333fred commented:
> I imagine this will also avoid the local being stacked in a ref assignment like this: `outParam1 = outParam2 = null;`, is that correct?

I'm not sure what you're asking @alrz: that isn't a ref assignment, this code should have no impact on it.

333fred commented:
@AlekseyTs addressed feedback

---
In reply to: [932432493](https://github.com/dotnet/roslyn/pull/56790#issuecomment-932432493) [](http://example.com/codeflow?ancestors=932432493,932432493)

cston commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -342,6 +342,10 @@
     <Field Name="Type" Type="TypeSymbol" Override="true" Null="disallow"/>
     <Field Name="Operand" Type="BoundExpression"/>
+    <!-- Used by emit to determine whether this pointer indirection is by value or by ref. If true, emit
+         will emit the address of the target, and not actually perform an indirect load (in in practice just
```
```suggestion
         will emit the address of the target, and not actually perform an indirect load (in practice just
```

AlekseyTs replied:
>in in [](http://example.com/codeflow?start=89&length=5)

Double "in"

cston commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -350,6 +354,10 @@
     <Field Name="Index" Type="BoundExpression"/>
     <Field Name="Checked" Type="bool"/>
+    <!-- Used by emit to determine whether this pointer element access is by value or by ref. If true, emit
+         will emit the address of the target, and not actually perform an indirect load (in in practice just
```
```suggestion
         will emit the address of the target, and not actually perform an indirect load (
         in in practice just
```

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -342,6 +342,10 @@
+    <!-- Used by emit to determine whether this pointer indirection is by value or by ref. If true, emit
+         will emit the address of the target, and not actually perform an indirect load (in in practice just
+         means emitting the Operand alone). -->
+    <Field Name="IsAssignmentAddress" Type="bool"/>
```
>IsAssignmentAddress [](http://example.com/codeflow?start=17&length=19)

Do we need the "Assignment" part in the name? Same for the other node.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs:
```diff
@@ -273,6 +273,22 @@ private BoundExpression CheckValue(BoundExpression expr, BindValueKind valueKind
+                    if ((valueKind & BindValueKind.RefersToLocation) == BindValueKind.RefersToLocation)
+                    {
+                        var pointerIndirection = (BoundPointerIndirectionOperator)expr;
+                        expr = pointerIndirection.Update(pointerIndirection.Operand, isAssignmentAddress: true, pointerIndirection.Type);
```
>isAssignmentAddress [](http://example.com/codeflow?start=85&length=19)

It feels like "onlyRefersToLocation" is a better name.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/Constructors.cs:
```diff
@@ -725,4 +725,24 @@ public BoundIncrementOperator Update(UnaryOperatorKind operatorKind, BoundExpres
+
+    internal partial class BoundPointerIndirectionOperator
+    {
+        public BoundPointerIndirectionOperator(SyntaxNode syntax, BoundExpression operand, TypeSymbol type, bool hasErrors = false)
```
>BoundPointerIndirectionOperator [](http://example.com/codeflow?start=15&length=31)

I think adding constructors like that is error prone, easy to miss the fact that a "non-default" value should be used. It also doesn't look like constructors are used very often, not that many changes are avoided. Same comment for the other bound node.

AlekseyTs commented:
Done with review pass (commit 7)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [932585188](https://github.com/dotnet/roslyn/pull/56790#issuecomment-932585188) [](http://example.com/codeflow?ancestors=932585188,932585188)

AlekseyTs reviewed (approved):
LGTM (commit 8), assuming CI is passing.
